docs/plugins/npe2_manifest_specification.md

```diff
@@ -0,0 +1,578 @@
+contributions:
+  commands:     <list of commands>
+  readers:      <list of readers>
+  writers:      <list of writers>
+  sample_data:  <list of sample data providers>
+  widgets:      <list of widget providers>
+  themes:       <list of themes>
+```
+
+Readers, writers, sample data providers and widget providers refer to callable
+python functions that a plugin defines. Each callable is identified with an
+entry in the list of `commands` via a unique id. For more see {ref}`Commands` below.
+
+```{note}
+Python package metadata (`setup.py` or `setup.cfg`) may be used to populate
+selected properties of the [PluginManifest][].
+```
+
+(top-level-props)=
+
+## Top-level properties
+
+### Required
+
+- **name** The name of the plugin. Example: `napari-svg`. Should be a
+  [PEP-8][]-compatible package name. If missing, this is populated from the
+  python package [name][setup-name].
+
+### Optional
+
+- **engine**: A SemVer compatible version string matching the versions of the
+  plugin engine that the extension is compatible with. Example: "0.1.0".
+- **display_name**: User-facing text to display as the name of this plugin.
+  Example: `napari SVG`. Must be 3-40 characters long, containing
+  printable word characters, and must not begin or end with an underscore,
+  white space, or non-word character.
+- **entry_point**: The module containing the `activate()` function. Example:
+  `foo.bar.baz`. This should be a fully qualified module string.
+
+(commands)=
+
+## Commands
+
+**command** is a python function associated with an id. The **id** is
+a unique identifier to which other contributions, like readers, can refer.
+
+### Required fields
+
+- **id** An identifier used to reference this command within this plugin.
+- **title** A description of the command. This might be used, for example,
+  when searching in a command palette. Examples: "Generate lily sample",
+  "Read tiff image", "Open gaussian blur widget".
+
+### Optional fields
+
+- **python_name** Fully qualified name to callable python object implementing
+  this command. This usually takes the form of
+  `{obj.__module__}:{obj.__qualname__}` (e.g.
+  `my_package.a_module:some_function`). For a command to be useful, this
+  field must be specified,. If it isn't specified here it should be done
+  dynamically inside the plugin's 'activate()' function.
+
+### Example
+
+```yaml
+name: my-plugin
+contribution:
+  commands:
+    - id: my-plugin.publish_paper
+      title: Do experiments, analysis, write paper, and submit
+      python_name: my_plugin:publish_func
+```
+
+### Dynamic command registration
+
+Commands can be registered dynamically by the plugin's `activate()` function.
+For example,
+
+```python
+# inside my_plugin/__init__.py
+from npe2 import PluginContext
+def activate(context: PluginContext):
+    @context.register_command("my_plugin.hello_world")
+    def _hello():
+        ...
+
+    context.register_command("my_plugin.another_command", lambda: print("yo!"))
```

Comment:
if I register a reader command like this, for example, do I still need to mention a reader contribution under "readers" in the manifest?

It's still not clear to me how to link these commands to the type of contribution they provide?

Reply:
yes, you'd still have to add a reader contribution.  If you were doing this fully dynamically you would (a) register the command (b) create and register a reader contribution.

If it's not clear that's because it's not clear :)

Reply:
I think what would make it clearer is to state that this pattern is an _alternative_ to using `python_name` in the command manifest.  That's all.  You still need all the other fields

Reply:
> state that this pattern is an alternative to using python_name in the command manifest

added that.  thanks!  Also turned this section into an admonition